### PR TITLE
[ENHANCEMENT] Add --force option to allow applying resource despite project config inconsistencies

### PR DESCRIPTION
**Pull request by ying-jeanne**
[ENHANCEMENT] Add --force option to allow applying resource despite project config inconsistencies

# Description
Fix: https://github.com/perses/perses/issues/1917, take a look at the problem, i am uncertain if this solution aligns with our objectives. The proposed PR addition of a force flag aims to alert users to potential project inconsistencies while still permitting them to create based on input metadata.

# Screenshots

# Checklist

- [x] Pull request has a descriptive title and context useful to a reviewer.
- [x] Pull request title follows the `[<catalog_entry>] <commit message>` naming convention using one of the
  following `catalog_entry` values: `FEATURE`, `ENHANCEMENT`, `BUGFIX`, `BREAKINGCHANGE`, `DOC`,`IGNORE`.
- [x] All commits have [DCO signoffs](https://github.com/probot/dco#how-it-works).

## UI Changes

- na Changes that impact the UI include screenshots and/or screencasts of the relevant changes.
- na Code follows the [UI guidelines](https://github.com/perses/perses/blob/main/ui/ui-guidelines.md).
- na Visual tests are stable and unlikely to be flaky.
  See [Storybook](https://github.com/perses/perses/tree/main/ui/storybook#visual-tests)
  and [e2e](https://github.com/perses/perses/tree/main/ui/e2e#visual-tests) docs for more details. Common issues
  include:
  - Is the data inconsistent? You need to mock API requests.
  - Does the time change? You need to use consistent time values or mock time utilities.
  - Does it have loading states? You need to wait for loading to complete.

### CODE REVIEW

**Review by celian-garcia — CHANGES_REQUESTED** [review]
Hello @ying-jeanne, I'm not sure yet that we want to implement it like that. The force flag if used should be consistent in its behaviour on all the commands and here we have a behaviour for apply only. 

I'd see at a glance that we should reuse --all which is always used for "do this command for all projects" it would probably fit better the use case. @Nexucis WDYT ?

**Comment by celian-garcia** [comment]
Also I think that an error message should be displayed at start of the command to prevent from applying half of the resources

**Comment by Nexucis** [comment]
Hello @ying-jeanne !

Thank you for taking the time to fix this issue :).

> I'd see at a glance that we should reuse --all which is always used for "do this command for all projects" it would probably fit better the use case. @Nexucis WDYT ?

usage of `--all` in the command `apply` would be weird.

Reading `percli apply -f data.json --all`, for me means I want to apply all data coming from the file `data.json`. Which would mean we are partially applying the resources from the file.

From my perspective, introducing a flag like `--force` is a nice way to solve it. If we don't like the meaning of `--force`, we could have instead the flag `--ignore.project-set`

> Also I think that an error message should be displayed at start of the command to prevent from applying half of the resources

That's something not possible currently as to be able to detect the inconsistency we need to loop other all resources that the user want to apply. 

The only way to do that, is to loop a first time to search this issue.

**Comment by celian-garcia** [comment]
I always see the ``force`` flag as something that you shouldn't do but as an unusual emergency case. Here it can be a valid daily case. To just to not have to manage your different projects separately.

I'm wondering after all if a warning wouldn't be sufficient.

Note for the overthinking buddies like me: Also the warnings in general could be hideable. Like if we had a common system of warnings in this case saying "project-inconsistency: you applied in the project 'hello' which is different from current selected project 'moto'" and a common --ignore flag that you could use to ignore that specific warning --ignore project-inconsistency, that would be good 
End of note for the overthinking buddies like me...

**Comment by ying-jeanne** [comment]
Hi @celian-garcia @Nexucis both, thanks for the review 💟 . I would make it draft for the moment, since I realised later yesterday besides the --froce flag, It would be better to apply this consistency check in validate() function instead and introduce a multi-errors in order to go through all resources and return all inconsistencies in the response just as you described above. Will continue later this week and we can continue the discussion afterwards.

**Comment by ying-jeanne** [comment]
I usually lean towards --force option since it's my go-to when I'm confident about my actions after reviewing what just happened/warned. However, I'm also receptive to considering --ignore.project-set. Personally, I don't have very strong preference here. In the updated code I've included the multierr functionality and relocated the validation logic to the validate() method. If you would like to review it again, just let me know wdyt.

File: internal/cli/cmd/apply/apply.go [hunk 1]
```diff
@@ -65,7 +68,30 @@ func (o *option) Complete(args []string) error {
 }
 
 func (o *option) Validate() error {
-	return nil
+	var result *multierror.Error
+	for _, entity := range o.entities {
+		kind := modelV1.Kind(entity.GetKind())
+		project := resource.GetProject(entity.GetMetadata(), o.Project)
+		if !o.ForceCreate && len(o.Project) != 0 && project != o.Project {
+			result = multierror.Append(result, fmt.Errorf("inconsistency has been detected for object %q %q: While the JSON metadata suggests the project name %q, the command-line indicates %q,", kind, entity.GetMetadata().GetName(), project, o.Project))
```
[the remark this answers is not in the document]
Comment (ying-jeanne) [reply internal/cli/cmd/apply/apply.go]: Hi this makes sense, updated PR accordingly

**Comment by celian-garcia** [comment]
Wow crazy I've been able to resolve the conflicts from Github UI directly despite the fact your PR comes from your fork :astonished: 
Github will always surprise me. <3

So @ying-jeanne thank you very much! Let's move forward and merge it, your changes are very nice.

File: go.mod [hunk 2]
```diff
@@ -15,6 +15,7 @@ require (
 	github.com/google/uuid v1.6.0
 	github.com/goreleaser/goreleaser v1.26.2
 	github.com/gorilla/securecookie v1.1.2
+	github.com/hashicorp/go-multierror v1.1.1
```
Comment (Nexucis) [inline go.mod:18]: Sorry for the long time it took me to go back at this PR.

The only thing I'm not ok with is the usage of library coming from hashicorp. (regarding the changes they made on their license), and besides introducing a new lib for managing an array of error, I find it a bit overkill.

File: internal/cli/opt/opt.go [hunk 3]
```diff
@@ -105,3 +105,11 @@ func (o *ProjectOption) Complete() error {
 func AddProjectFlags(cmd *cobra.Command, o *ProjectOption) {
 	cmd.Flags().StringVarP(&o.Project, "project", "p", o.Project, "If present, the project scope for this CLI request")
 }
+
+type ForceCreateOption struct {
```
Comment (Nexucis) [inline internal/cli/opt/opt.go:109]: Do you see an usage of this flag in other command ? If no, probably it's not worth it to create a struct to handle this flag

**Comment by ying-jeanne** [comment]
> Wow crazy I've been able to resolve the conflicts from Github UI directly despite the fact your PR comes from your fork 😲 Github will always surprise me. <3
> 
> So @ying-jeanne thank you very much! Let's move forward and merge it, your changes are very nice.

I signed share my fork with project maintainer! This is nice because back and force is so painful <3 thank you for the review @celian-garcia

File: go.mod (hunk 2 above)
Comment (ying-jeanne) [reply go.mod:18]: I hesitated creating a helper multierror in the project directly, look at go-multierror codebase, it is light with very few updates/CVEs. I understand each projects have their own preference, if you prefer less dependency, I could replace it.

File: internal/cli/cmd/apply/apply.go [hunk 4]
```diff
@@ -17,6 +17,7 @@ import (
 	"fmt"
 	"io"
 
+	"github.com/efficientgo/core/merrors"
```
Comment (ying-jeanne) [inline internal/cli/cmd/apply/apply.go:20]: while doing something in prometheus go client, found this lib wrote by our dear bartek, efficient and lighter than multierror

File: internal/cli/cmd/apply/apply.go [hunk 5]
```diff
@@ -65,7 +68,30 @@ func (o *option) Complete(args []string) error {
 }
 
 func (o *option) Validate() error {
-	return nil
+	var result *multierror.Error
```
[the remark this answers is not in the document]
Comment (ying-jeanne) [reply internal/cli/cmd/apply/apply.go]: changed in 7c25ff5396abc6515bf8fef305b78c5312981481

File: internal/cli/opt/opt.go (hunk 3 above)
Comment (ying-jeanne) [reply internal/cli/opt/opt.go]: removed in 7c25ff5396abc6515bf8fef305b78c5312981481

File: internal/cli/cmd/apply/apply.go [hunk 6]
```diff
@@ -65,7 +67,23 @@ func (o *option) Complete(args []string) error {
 	return o.setEntities()
 }
 
+func (o *option) validateProjectConsistency() error {
+	var result = merrors.New()
+	for _, entity := range o.entities {
+		kind := modelV1.Kind(entity.GetKind())
+		project := resource.GetProject(entity.GetMetadata(), o.Project)
+		if !o.forceCreate && len(o.Project) != 0 && project != o.Project {
```
Comment (celian-garcia) [inline internal/cli/cmd/apply/apply.go:75]: You can remove the !o.forceCreate condition from there.

**Review by celian-garcia — APPROVED** [review]
Thanks @ying-jeanne for reviving this! Just added a nit comment but not very important. 
I'm personally ok with this new lib, but I'd wait for the @Nexucis opinion as it was his concern.

**Comment by Nexucis** [comment]
Looks better indeed.
Thank you for your permanent work on this PR @ying-jeanne and sorry for the long delay it took to review / gives opinion.

Once last comment from @celian-garcia are tackled I am fine to merge it.

**Comment by Nexucis** [comment]
if by any chance you pass here to fix it soon @ying-jeanne , would be cool to merge it before making the new release !

**Comment by ying-jeanne** [comment]
comments addressed in commit [8d2f589](https://github.com/perses/perses/pull/1937/commits/8d2f589c45771cd4731912a59c5cc1190c190d8f) (:

**Review by Nexucis — APPROVED** [review]
awesome thanks !

**Pull request merged**